src/Agent.Plugins/PipelineCache/FingerprintCreator.cs

```diff
@@ -0,0 +1,244 @@
+                enumerateRootPath = workingDirectory;
+                enumeratePattern = "*";
+                enumerateDepth = SearchOption.AllDirectories;
+            }
+            else
+            {
+                int rootDirLength = rootRule.Substring(0,firstGlob).LastIndexOfAny( new [] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar});
+                enumerateRootPath = rootRule.Substring(0,rootDirLength);
+                enumeratePattern = "*";
+                enumerateDepth = SearchOption.AllDirectories;
+            }
+        }
+
+        public static Fingerprint ParseFromYAML(
+            AgentTaskPluginExecutionContext context,
+            IEnumerable<string> keySegments,
+            bool addWildcard)
+        {
+            var sha256 = new SHA256Managed();
+
+            string workingDirectoryValue = context.Variables.GetValueOrDefault(
+                "system.defaultworkingdirectory" // Constants.Variables.System.DefaultWorkingDirectory
+                )?.Value;
+
+            var resolvedSegments = new List<string>();
+
+            foreach (string keySegment in keySegments)
+            {
+                if (keySegment.Length == 1 && keySegment[0] == '*')
+                {
+                    throw new ArgumentException("`*` is a reserved key segment. For path glob, use `./*`.");
+                }
+                else if (keySegment.Equals(Fingerprint.Wildcard, StringComparison.Ordinal))
+                {
+                    throw new ArgumentException("`**` is a reserved key segment. For path glob, use `./**`.");
+                }
+                else if (IsPathy(keySegment))
+                {
+                    context.Verbose($"Interpretting `{keySegment}` as a path.");
+
+                    var segment = new StringBuilder();
+                    bool foundFile = false;
+
+                    if (keySegment.Contains(';', StringComparison.Ordinal))
+                    {
+                        throw new ArgumentException("Cache key cannot contain the ';' character.");
+                    }
+
+                    string[] pathRules = keySegment.Split(new []{','}, StringSplitOptions.RemoveEmptyEntries);
+                    string rootRule = pathRules.First();
+                    if(rootRule.Length == 0 || rootRule[1] == '!')
+                    {
+                        throw new ArgumentException();
+                    }
+
+                    string workingDirectory = null;
+                    if (!IsAbsolutePath(rootRule))
+                    {
+                        workingDirectory = workingDirectoryValue;
+                    }
+
+                    string absoluteRootRule = MakePathAbsolute(workingDirectory, rootRule);
+                    context.Verbose($"Expanded include rule is `{absoluteRootRule}`.");
+                    IEnumerable<string> absoluteExcludeRules = pathRules.Skip(1).Select(r => MakePathAbsolute(workingDirectory, r.Substring(1)));
```

Comment:
Let's think about whether we should be stricter (for now) and **not allow absolute file paths in "pathy" key segments that use wildcards**, but still allow them in non-wildcard segments. This reduces the surface area of where we have to look for matching files. It also lets the user specify multiple "inclusions" rules in the same key segment, which can be followed by zero or more "exclusions" rules. Every rule would be rooted at `$(System.DefaultWorkingDirectory)`.

The enumeration path would always be at least `$(System.DefaultWorkingDirectory)`, but if all the inclusion rules were, for example, under a different path (e.g. `packages/*.lock, packages/package.json`), the enumeration path could be just `$(System.DefaultWorkingDirectory)/packages`.

Reply:
Haha! I had cut that because I though you didn't want it! 

We could allow multiple includes and multiple excludes within a given key segment, but it could get ugly :)